staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go

```diff
@@ -47,12 +50,22 @@ type customResourceStrategy struct {
 	namespaceScoped   bool
 	validator         customResourceValidator
 	structuralSchemas map[string]*structuralschema.Structural
+	celValidators     map[string]*cel.Validator
 	status            *apiextensions.CustomResourceSubresourceStatus
 	scale             *apiextensions.CustomResourceSubresourceScale
 	kind              schema.GroupVersionKind
 }
 
 func NewStrategy(typer runtime.ObjectTyper, namespaceScoped bool, kind schema.GroupVersionKind, schemaValidator, statusSchemaValidator *validate.SchemaValidator, structuralSchemas map[string]*structuralschema.Structural, status *apiextensions.CustomResourceSubresourceStatus, scale *apiextensions.CustomResourceSubresourceScale) customResourceStrategy {
+	var celValidators map[string]*cel.Validator
+	if utilfeature.DefaultFeatureGate.Enabled(features.CustomResourceValidationExpressions) {
+		celValidators = make(map[string]*cel.Validator, len(structuralSchemas))
+		for name, s := range structuralSchemas {
+			v := cel.NewValidator(s) // CEL programs are compiled and cached here
```

Comment:
if s contains no cel expressions, should v be nil?

Reply:
Makes sense. Now we only have validators when we need them, and only the minimal overlay of validators that are needed is constructed.